### PR TITLE
Add scalar/Neon/Neon hybrid for Keccak-x4

### PR DESCRIPTION
**Pull request by hanno-becker**
This PR adds hybrid scalar/Neon/Neon implementations of Keccak-x4. Those implementations were first described in https://eprint.iacr.org/2022/1243.

We don't consume the implementations of that paper as-is, however, but auto-generate the code from clean de-interleaved assembly. As it stands, only the interleaved versions are added here.

### CODE REVIEW

**Comment by hanno-becker** [comment]
I haven't yet amended `optimize.sh` to reproduce the optimizations from de-interleaved code. As it stands, the optimizations are only implemented in `example.py` in the draft PR https://github.com/slothy-optimizer/slothy/pull/65. 

I'd prefer for this not to block the PR, but could be convinced otherwise.

**Review by mkannwischer — APPROVED** [review]
Thanks @hanno-becker!!
It's getting rather hard to keep track of which implementation is used for each target. #141 should be addressed soon. 

Code looks mostly good to me - some comments. 
Here is the performance of `keccak-f1600-x4 cycles` 
| platform | before | after  | 
| ------------- | --------- | -------- |
| A72          | 6196    | 5407  |
| A76          | 3708    | 2607  | 
| A55          | 6012    | ~6346~  | 
| G2            | 3709    | 2608  | 
| G3            | 2078    | 1604  | 

~That looks great except for the A55 - there, you just want to use a scalar 1x implementation throughout. That's in line with the paper. 
We'll need a special case for the A55 I guess. Probably better to do that as a follow-up PR.~

Thanks for introducing the profile for the special case A55.

File: fips202/keccakf1600.c [hunk 1]
```diff
@@ -110,16 +110,51 @@ static void keccakx2_unzip(uint64_t *state)
 
     memcpy(state, tmp, sizeof(tmp));
 }
-#endif /* MLKEM_USE_FIPS202_X2_ASM && MLKEM_USE_FIPS202_X2_ASM_ZIPPED */
+#endif
+
+#if defined(MLKEM_USE_FIPS202_X4_ASM) && defined(MLKEM_USE_FIPS202_X4_ASM_ZIPPED)
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/keccakf1600.c:115]: At the moment, they aren't used. The implementations we developed in the SPHINCS+ paper were all zipped, though.

File: fips202/asm/aarch64/keccak_f1600_x4_v8a_v84a_scalar_hybrid.S [hunk 2]
```diff
@@ -0,0 +1,1660 @@
...
+#define STACK_OFFSET_CONST_VECTOR (2*8)
+#define STACK_OFFSET_COUNT (3*8)
+#define STACK_OFFSET_OUTER (4*8)
+
+#define STACK_LOC_0 ((STACK_OFFSET_LOCS) + 0*8)
+#define STACK_LOC_1 ((STACK_OFFSET_LOCS) + 1*8)
+#define STACK_LOC_2 ((STACK_OFFSET_LOCS) + 2*8)
+#define STACK_LOC_3 ((STACK_OFFSET_LOCS) + 3*8)
+#define STACK_LOC_4 ((STACK_OFFSET_LOCS) + 4*8)
+#define STACK_LOC_5 ((STACK_OFFSET_LOCS) + 4*8)
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/asm/aarch64/keccak_f1600_x4_v8a_v84a_scalar_hybrid.S]: Fixed

File: fips202/asm/aarch64/keccak_f1600_x4_v8a_scalar_hybrid.S [hunk 3]
```diff
@@ -0,0 +1,1735 @@
...
+#define STACK_OFFSET_CONST_VECTOR (2*8)
+#define STACK_OFFSET_COUNT (3*8)
+#define STACK_OFFSET_OUTER (4*8)
+
+#define STACK_LOC_0 ((STACK_OFFSET_LOCS) + 0*8)
+#define STACK_LOC_1 ((STACK_OFFSET_LOCS) + 1*8)
+#define STACK_LOC_2 ((STACK_OFFSET_LOCS) + 2*8)
+#define STACK_LOC_3 ((STACK_OFFSET_LOCS) + 3*8)
+#define STACK_LOC_4 ((STACK_OFFSET_LOCS) + 4*8)
+#define STACK_LOC_5 ((STACK_OFFSET_LOCS) + 4*8)
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/asm/aarch64/keccak_f1600_x4_v8a_scalar_hybrid.S]: Fixed

File: fips202/asm/aarch64/keccak_f1600_x4_v84a_scalar_hybrid.S [hunk 4]
```diff
@@ -0,0 +1,1570 @@
...
+#define STACK_OFFSET_CONST_VECTOR (2*8)
+#define STACK_OFFSET_COUNT (3*8)
+#define STACK_OFFSET_OUTER (4*8)
+
+#define STACK_LOC_0 ((STACK_OFFSET_LOCS) + 0*8)
+#define STACK_LOC_1 ((STACK_OFFSET_LOCS) + 1*8)
+#define STACK_LOC_2 ((STACK_OFFSET_LOCS) + 2*8)
+#define STACK_LOC_3 ((STACK_OFFSET_LOCS) + 3*8)
+#define STACK_LOC_4 ((STACK_OFFSET_LOCS) + 4*8)
+#define STACK_LOC_5 ((STACK_OFFSET_LOCS) + 4*8)
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/asm/aarch64/keccak_f1600_x4_v84a_scalar_hybrid.S]: Fixed (need to keep 2 for alignment)

File: fips202/asm/aarch64/keccak_f1600_x4_v84a_scalar_hybrid.S
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/asm/aarch64/keccak_f1600_x4_v84a_scalar_hybrid.S:1]: done

**Comment by mkannwischer** [comment]
Thanks! Now A55 performance looks good again, too. 
LGTM.

**Pull request merged**